### PR TITLE
SeqOut first implementation

### PR DESCRIPTION
**Pull request by hy0000**
see SeqOut and SeqOutTest

### CODE REVIEW

File: rocev2/test/src/scala/rdma/RecvQTest.scala [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+    ) {
+      timeOut -= 1
+      dut.clockDomain.waitRisingEdge()
+      if (timeOut == 0) {
+        simFailure(
+          Seq(
+            s"sim time out when waiting psn = ${psn}",
+            s"rxOtherResp.psn = ${dut.io.rxOtherResp.bth.psn.toInt}, valid ${dut.io.rxOtherResp.valid.toBoolean}, ready ${dut.io.rxOtherResp.ready.toBoolean}",
+            s"rxReadResp.psn = ${dut.io.rxReadResp.bth.psn.toInt}, valid ${dut.io.rxReadResp.valid.toBoolean}, ready ${dut.io.rxReadResp.ready.toBoolean}",
+            s"tx.psn          = ${dut.io.tx.bth.psn.toInt}, valid ${dut.io.tx.valid.toBoolean}, ready ${dut.io.tx.ready.toBoolean}"
```
Comment (pwang7) [inline rocev2/test/src/scala/rdma/RecvQTest.scala:143]: align to =

File: rocev2/test/src/scala/rdma/RecvQTest.scala [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+        val r = Random.nextInt(distance)
+        val j = i + r
+        if (j < psnList.length && !moved(i)) {
+          val t = psnList(j)
+          psnList = psnList.updated(j, psnList(i))
+          psnList = psnList.updated(i, t)
+          moved = moved.updated(j, true)
+        }
+      }
+      // println(s"generate psn from $PSN until ${PSN + n}")
```
Comment (pwang7) [inline rocev2/test/src/scala/rdma/RecvQTest.scala:100]: remove commented code

File: rocev2/src/main/scala/rdma/RecvQ.scala [hunk 3]
```diff
@@ -188,20 +193,82 @@ class SeqOut(busWidth: BusWidth) extends Component {
...
+    // avoid conflict when rxOtherResp try to set it True
+    camInvalidate.data := writePsn === invalidatePsn && io.rxOtherResp.fire
+    camInvalidate.valid := RegNext(psnIncrease)
+
+    if (bootFlush) {
+      val flushed = Reg(True)
+      val flushCounter = Counter(bufferDepth, !flushed)
+      flushed.setWhen(flushCounter.willOverflow)
+      when(!flushed) {
+        //TODO:flush logic
```
Comment (pwang7) [inline rocev2/src/main/scala/rdma/RecvQ.scala:225]: // TODO: flush logic

**Review by pwang7 — CHANGES_REQUESTED** [review]

File: rocev2/src/main/scala/rdma/RecvQ.scala [hunk 4]
```diff
@@ -324,28 +325,50 @@ class SeqOut(busWidth: BusWidth) extends Component {
...
     assert(
       assertion = OpCode.isRespPkt(txSel.bth.opcode),
       message =
         L"SeqOut can only output response packet, but with invalid opcode=${txSel.bth.opcode}",
       severity = ERROR
     )
+
+    assert(
+      assertion = CountOne(rxFiltered.map(_.valid)) <= 1,
+      message = "SeqOut has duplicate PSN input",
```
Comment (pwang7) [inline rocev2/src/main/scala/rdma/RecvQ.scala:361]: message = L"SeqOut has duplicate PSN input, oPSN=${opsnReg}",

File: rocev2/src/main/scala/rdma/RecvQ.scala [hunk 5]
```diff
@@ -324,28 +325,50 @@ class SeqOut(busWidth: BusWidth) extends Component {
...
+      severity = ERROR
+    )
+  }
+
+  val rxAllValid = rxSeq.map(_.valid).reduce(_ && _)
```
Comment (pwang7) [inline rocev2/src/main/scala/rdma/RecvQ.scala:366]: val rxAllValid = Vec(rxSeq.map(_.valid)).asBits.andR

File: rocev2/test/src/scala/rdma/RecvQTest.scala [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+  }
+
+  def simSeqOut(dut: SeqOut): Unit = {
+    dut.clockDomain.forkStimulus(2)
+    val (rxDrivers, psnSetter, targetPsnWaiter) = setInputDrivers(dut)
+    val psnGen = new PsnGenerator
+    val maxPSN = (1 << PSN_WIDTH) - 1
+    val maxPacketNum = 0xff
+
+    // test 1, case for psn wrap
```
Comment (pwang7) [inline rocev2/test/src/scala/rdma/RecvQTest.scala:154]: separate to different test cases

File: rocev2/test/src/scala/rdma/RecvQTest.scala [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    psnSetter.set(maxPSN - 10)
+    psnGen.setPsn(maxPSN - 10)
+    val psnList1 = (maxPSN - 10 to maxPSN).toList
+    val psnList2 = (0 to 10).toList
+    rxDrivers(1).push(psnList1)
+    rxDrivers(2).push(psnList2)
+    targetPsnWaiter.add(psnList1 ++ psnList2)
+    targetPsnWaiter.start()
+
+    // test 2, random test
```
Comment (pwang7) [inline rocev2/test/src/scala/rdma/RecvQTest.scala:164]: separate to different test cases

File: rocev2/test/src/scala/rdma/RecvQTest.scala [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    }
+  }
+
+  class TargetPsnWaiter(dut: SeqOut) {
+    var psnAll: List[Int] = List()
+
+    fork {
+      while (true) {
+        dut.io.tx.ready #= !dut.io.tx.ready.toBoolean
+        dut.clockDomain.waitRisingEdge(Random.nextInt(8))
```
Comment (pwang7) [inline rocev2/test/src/scala/rdma/RecvQTest.scala:86]: why nextInt(8)? need to add comments to all constant numbers

File: rocev2/test/src/scala/rdma/RecvQTest.scala [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+    val rxSendDriver =
+      new RxDriver(dut.io.rxSendResp, dut.clockDomain, OpCode.ACKNOWLEDGE, 1)
+    val rxAtomicDriver = new RxDriver(
+      dut.io.rxAtomicResp,
+      dut.clockDomain,
+      OpCode.ATOMIC_ACKNOWLEDGE,
+      1
+    )
+    val rxWriteDriver =
+      new RxDriver(dut.io.rxWriteResp, dut.clockDomain, OpCode.ACKNOWLEDGE, 1)
```
Comment (pwang7) [inline rocev2/test/src/scala/rdma/RecvQTest.scala:125]: do not use constant literal numbers directly, define a val instead

File: rocev2/test/src/scala/rdma/RecvQTest.scala [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+    rxDrivers(1).push(psnList1)
+    rxDrivers(2).push(psnList2)
+    targetPsnWaiter.add(psnList1 ++ psnList2)
+    targetPsnWaiter.start()
+
+    // test 2, random test
+    for (psn <- Seq.fill(64)(Random.nextInt(maxPSN - maxPacketNum))) {
+      psnSetter.set(psn)
+      psnGen.setPsn(psn)
+      for (_ <- 0 to 8) {
```
Comment (pwang7) [inline rocev2/test/src/scala/rdma/RecvQTest.scala:168]: do not use 8, use rxDrivers.length instead

File: rocev2/test/src/scala/rdma/RecvQTest.scala [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+    def set(psn: Int): Unit = {
+      qpAttr.epsn #= psn
+      qpAttrUpdate.pulseRqPsnReset #= true
+      clockDomain.waitRisingEdge()
+      qpAttrUpdate.pulseRqPsnReset #= false
+    }
+  }
+
+  class PsnGenerator {
+    var PSN: Int = -1
```
Comment (pwang7) [inline rocev2/test/src/scala/rdma/RecvQTest.scala:67]: `val psn`, use camel case style,
names with all capital letters represent constants,
do not use var.

File: rocev2/test/src/scala/rdma/RecvQTest.scala [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+    def setPsn(psn: Int): this.type = {
+      PSN = psn
+      this
+    }
+
+    def genOrderedNext(n: Int): List[Int] = {
```
Comment (pwang7) [inline rocev2/test/src/scala/rdma/RecvQTest.scala:73]: better to generate a sequence of PSN for test

File: rocev2/test/src/scala/rdma/RecvQTest.scala [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+      OpCode.ATOMIC_ACKNOWLEDGE,
+      1
+    )
+    val rxWriteDriver =
+      new RxDriver(dut.io.rxWriteResp, dut.clockDomain, OpCode.ACKNOWLEDGE, 1)
+    val rxReadDriver = new RxDriver(
+      dut.io.rxReadResp,
+      dut.clockDomain,
+      OpCode.RDMA_READ_RESPONSE_ONLY,
+      8
```
Comment (pwang7) [inline rocev2/test/src/scala/rdma/RecvQTest.scala:130]: do not use constant number directly

**Review by pwang7 — CHANGES_REQUESTED** [review]

File: rocev2/test/src/scala/rdma/RecvQTest.scala [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+      driver.push(psnList)
+      targetPsnWaiter.wait(psnList)
+    })
+  }
+
+  def simSeqOutRandom(dut: SeqOut): Unit = {
+    dut.clockDomain.forkStimulus(2)
+    val (rxDrivers, psnSetter, targetPsnWaiter) = setInputDrivers(dut)
+    val maxPSN = (1 << PSN_WIDTH) - 1
+    val maxPacketsNum = 0xff
```
Comment (pwang7) [inline rocev2/test/src/scala/rdma/RecvQTest.scala:164]: usually hex is used for mask, use decimal for number

File: rocev2/test/src/scala/rdma/RecvQTest.scala [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+
+    def push(psnList: List[Int]): Unit = {
+      psnQueue ++= psnList
+    }
+
+    def clear(): Unit = {
+      psnQueue.clear()
+    }
+
+    val driver = fork {
```
Comment (pwang7) [inline rocev2/test/src/scala/rdma/RecvQTest.scala:33]: use def?

File: rocev2/test/src/scala/rdma/RecvQTest.scala [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+    val rxDrivers = Seq(
+      rxAtomicDriver,
+      rxReadDriver
+    ) ++ Seq(
+      dut.io.rxErrReqResp,
+      dut.io.rxDupReqResp,
+      dut.io.rxSendResp
+    ).map(
+      new RxDriver(
+        _,
```
Comment (pwang7) [inline rocev2/test/src/scala/rdma/RecvQTest.scala:131]: use explicit val for better readability

File: rocev2/test/src/scala/rdma/RecvQTest.scala [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+          }
+        } else {
+          rx.valid #= false
+          clockDomain.waitRisingEdge()
+        }
+      }
+    }
+  }
+
+  class PsnSetter(
```
Comment (pwang7) [inline rocev2/test/src/scala/rdma/RecvQTest.scala:53]: this class has no member fields, consider refactor it to object

**Review by pwang7 — CHANGES_REQUESTED** [review]

File: rocev2/test/src/scala/rdma/RecvQTest.scala (hunk 15 above)
Comment (hy0000) [reply rocev2/test/src/scala/rdma/RecvQTest.scala:33]: this fork block auto start when the class be created, I don't mean to call it manually. so let me delete the assign from fork to a val.

File: rocev2/test/src/scala/rdma/RecvQTest.scala (hunk 17 above)
Comment (hy0000) [reply rocev2/test/src/scala/rdma/RecvQTest.scala:53]: It's a driver bind with io bundles. I think the class semantic is more suitable this case, I found that other io binders are define as class in SpinalHDL source.

**Review by pwang7 — APPROVED** [review]

**Pull request merged**